[CodeCompletion][Sema][Parse] Migrate unresolved member completion to the solver-based completion implementation

nathawes opened the pull request:
Following on from updating regular member completion, this hooks up unresolved member completion (i.e. `.<complete here>`) to the `typeCheckForCodeCompletion` API to generate completions from all solutions the constraint solver produces (even those requiring fixes), rather than relying on a single solution being applied to the AST (if any). This lets us produce unresolved member completions even when the contextual type is ambiguous or involves errors.

Whenever `typeCheckExpression` is called on an expression containing a code completion expression and a `TypeCheckCompletionCallback` has been set, each solution formed is passed to the callback so the type of the completion expression can be extracted and used to lookup up the members to return.

Resolves rdar://problem/71378252

Commit:
[CodeCompletion][Sema][Parse] Migrate unresolved member completion to the solver-based completion implementation

Following on from updating regular member completion, this hooks up unresolved
member completion (i.e. .<complete here>) to the typeCheckForCodeCompletion API
to generate completions from all solutions the constraint solver produces (even
those requiring fixes), rather than relying on a single solution being applied
to the AST (if any). This lets us produce unresolved member completions even
when the contextual type is ambiguous or involves errors.

Whenever typeCheckExpression is called on an expression containing a code
completion expression and a CompletionCallback has been set, each solution
formed is passed to the callback so the type of the completion expression can
be extracted and used to lookup up the members to return.

nathawes commented:
@swift-ci please test

nathawes commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3581,26 +3581,22 @@ Parser::parseExprCollectionElement(Optional<bool> &isDictionary) {
-                               new (Context) CodeCompletionExpr(PreviousLoc));
-    }
+    Value = makeParserResult(makeParserError(),
+                             new (Context) ErrorExpr(SourceRange()));
```
@rintaro is making an ErrorExpr with an invalid source range for a "missing" value ok? This was needed for test cases like `let x: [EnumA:String] = [.a: "foo", foo.<complete>]` so that constraint generation treats the last element as a key. I don't have any appropriate source location we can give it. The PreviousLoc is the key, so if we used that the value's source range ends up overlapping the key's which seems bad, and the loc after the key may be EOF, ], or something random. The other option is to just put the element in as is and update either constraint generation or solving (like we did for dictionary/array literal mismatch), but I'm not sure how many places assume the elements of DictionaryExpr are TupleExprs.

rintaro replied:
I *think* we can use a invalid source range here. As far as I remember, in code completion, nothing uses the range of the `ErrorExpr`. Also `TupleExpr::getSourceRange()` correctly ignores invalid element source ranges so diagnostics shouldn't be affected.

xedin replied:
Do you think it would make sense to introduce a special `HoleExpr` as a placeholder for missing AST elements just like we introduce `HoleType` into the constraint system to indicate that something couldn't be inferred?

nathawes commented on test/IDE/complete_ambiguous.swift:
```diff
@@ -117,13 +119,33 @@ let x: A = overloadedReturn(1).#^RELATED_EXTRAARG^#
+// UNRESOLVED_UNAMBIGUOUS: Begin completions, 2 items
+// UNRESOLVED_UNAMBIGUOUS-DAG: Decl[InstanceMethod]/CurrNominal/TypeRelation[Invalid]: doBThings({#(self): B#})[#() -> Void#]{{; name=.+$}}
+// UNRESOLVED_UNAMBIGUOUS-DAG: Decl[Constructor]/CurrNominal/TypeRelation[Identical]: init()[#B#]{{; name=.+$}}
+// UNRESOLVED_UNAMBIGUOUS: End completions
```
I just realized this is probably undesirable behavior on the part of the filtering logic 😞. We get a viable solution choosing the `overloadedArg(_: B) -> B` overload and an unviable one choosing `overloadedArg(_: A) -> A` (because of arg/param mismatch) that gets filtered out. For code completions purposes we should probably ignore arg/param mismatches when the arg contains the code completion loc as it may be incomplete. E.g. the user may intend to do `overloadedArg(.somethingThatReturnsA()).somethingThatReturnsB()` making the `overloadedArg(_: A) -> A` overload viable as an arg for `takesB` again.

nathawes replied:
Fixed.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

rintaro commented on test/IDE/complete_unresolved_members.swift:
```diff
@@ -28,8 +28,8 @@
 // RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=UNRESOLVED_19 | %FileCheck %s -check-prefix=UNRESOLVED_1
```
I guess this patch fixes (currently disabled) `UNRESOLVED_44`.

nathawes replied:
Oh, good catch - it does. Let me re-enable it.

Commit:
[test] Re-enable now-passing IDE/complete_unresolved_members.swift run line

nathawes commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3802,6 +3802,24 @@ bool ConstraintSystem::repairFailures(
+        bool isSynthesizedArg = arg->isImplicit() && isa<DeclRefExpr>(arg);
+        SourceRange range = arg->getSourceRange();
+        if (!isSynthesizedArg && range.isValid() &&
+            Context.SourceMgr.rangeContainsCodeCompletionLoc(range) &&
```
I think it would be good to get `range.isValid() &&
            Context.SourceMgr.rangeContainsCodeCompletionLoc(range)` condition extracted into a separate method on constraint system, something like `ConstraintSystem::containsCodeCompletionToken(Expr *expr)` since that's a pattern we here and in CSGen.

xedin reviewed (approved):
Type checker related changes LGTM! I have left a small comment inline.

rintaro commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1153,31 +1203,45 @@ sawSolution(const constraints::Solution &S) {
-    } else if (auto *ParentCE = dyn_cast<ClosureExpr>(ParentExpr)) {
-      if (ParentCE->hasSingleExpressionBody() &&
-          ParentCE->getSingleExpressionBody() == CompletionExpr) {
-        ASTNode Last = ParentCE->getBody()->getLastElement();
-        if (!Last.isStmt(StmtKind::Return) || Last.isImplicit())
-          SingleExprBody = true;
```
Could you remind me why this check is needed (why `return (ParentCE->hasSingleExpressionBody() &&
           ParentCE->getSingleExpressionBody() == CompletionExpr)` is not enough)?

rintaro replied:
I *think* this effectively only checks the return statement is `isImplicit()`. Why we don't treat `{ return foo.<HERE> }` a single expression closure, although the typechecker does?

nathawes replied:
I think this was to match the existing behavior before the solver migration. From memory when the 'isSingleExpressionBody' bool is true the type relation computation avoids penalizing results that don't match the expected type, but still prioritizes ones that do.  I think the reasoning is that the user may intend to write multiple statements, but just hasn't yet.  E.g. if the user intends to write the below
```
someCallToClosureReturningBool {
   print("hello")
   return true
 }
```
and uses code completion at this point:
```
someCallToClosureReturningBool {
   <complete>
 }
```
We should still prioritize results returning bool (in case they really do intend to write a single expression closure), but shouldn't demote "print", for example, as they may add a separate return statement later. If they explicitly wrote a return statement though, like below, we should demote "print":
```
someCallToClosureReturningBool {
   return <complete>
 }
```
I guess it would be more accurate to name it `isImplicitSingleExpressionReturn` or maybe `ignoreExpectedTyMismatch`. Do you think one of those (or something else) would be better?

nathawes replied:
The ignoring invalid types when isSingleExpressionBody is true happens here: https://github.com/apple/swift/blob/main/lib/IDE/CodeCompletion.cpp#L1181

rintaro replied:
Thanks for the explanation! `isImplicitSingleExpressionReturn` looks better I think. Also, could you add some comment?

rintaro reviewed (approved):
Completion changes look good. Thank you!

Commit:
[Sema] Extract out expression-contains-completion-loc check into a method on ConstraintSystem (NFC)

Commit:
[CodeCompletion] Rename isSingleExpressionBody to isImplicitSingleExpressionReturn for clarity

This bool was true only for single-expression closure and function bodies that
do not have an explicit `return` in the source. Rename it and its associated
methods, to avoid confusion with the hasSingleExpressionBody methods on
ClosureExpr and AbstractFuncDecl. Those don't care whether the expression was
explicitly written in the source or not.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test Windows Platform
